File: packages/simulation-core/Sources/SimulationCore/PitchSequenceEvaluator.swift
import Foundation

/// A positive strategic pattern recognized after a pitch has resolved.
///
/// These tags describe the player's choices. They are deliberately downstream of
/// `PitchKernelEngine`, so recognizing one can never change pitch execution or outcome odds.
public enum PitchSequenceTag: String, Codable, CaseIterable, Sendable {
    case speedLadder = "speed_ladder"
    case eyeLevelChange = "eye_level_change"
    case insideOutside = "inside_outside"
    case expandAfterTwoStrikes = "expand_after_two_strikes"
    case stealStrike = "steal_strike"
    case counterRead = "counter_read"
}

/// The choice and result data needed to recognize a pitch sequence.
///
/// `expectedVelocityKPH` describes the called pitch, rather than the randomized measured
/// velocity. This keeps recognition tied to player intent and deterministic across presentation
/// layers.
public struct PitchSequencePitch: Codable, Equatable, Sendable {
    public let pitchType: PitchType
    public let zone: PitchZone
    public let intent: ZoneIntent
    public let expectedVelocityKPH: Int
    public let outcome: PitchOutcome

    public init(
        pitchType: PitchType,
        zone: PitchZone,
        intent: ZoneIntent,
        expectedVelocityKPH: Int,
        outcome: PitchOutcome
    ) {
        self.pitchType = pitchType
        self.zone = zone
        self.intent = intent
        self.expectedVelocityKPH = expectedVelocityKPH
        self.outcome = outcome
    }
}

/// One player-facing mastery badge. The evaluator returns at most one moment per pitch.
public struct PitchSequenceMoment: Codable, Equatable, Sendable {
    public let pitchNumber: Int
    public let tag: PitchSequenceTag
    public let headline: String
    public let detail: String

    public init(
        pitchNumber: Int,
        tag: PitchSequenceTag,
        headline: String,
        detail: String
    ) {
        self.pitchNumber = pitchNumber
        self.tag = tag
        self.headline = headline
        self.detail = detail
    }
}

/// Pure, deterministic recognition layered on top of an already-resolved pitch.
public enum PitchSequenceEvaluator {
    /// A visible change of at least this size is large enough to call a deliberate speed ladder.
    public static let minimumSpeedDifferenceKPH = 12

    /// Evaluates `current` against up to the last three earlier pitches.
    ///
    /// - Important: `recent` must not contain `current`. `rivalMemory` is the memory visible before
    ///   `current` was thrown, so `counter_read` only celebrates reacting to an existing warning.
    /// - Returns: At most one moment. More situation-specific reads take priority over generic
    ///   vertical, horizontal, and velocity changes.
    public static func evaluate(
        recent: [PitchSequencePitch],
        context: PlateAppearanceContext,
        current: PitchSequencePitch,
        rivalMemory: RivalMemorySnapshot?
    ) -> PitchSequenceMoment? {
        guard isValid(current),
              (0...3).contains(context.balls),
              (0...2).contains(context.strikes),
              context.pitchNumber > 0 else {
            return nil
        }

        let recentWindow = recent.suffix(3)

        if recognizesCounterRead(
            context: context,
            current: current,
            rivalMemory: rivalMemory
        ) {
            return moment(
                context: context,
                tag: .counterRead,
                headline: "읽힘을 역이용했다",
                detail: "상대 벤치가 읽은 반복을 끊고 좋은 결과를 만들었습니다."
            )
        }

        if context.strikes == 2,
           current.intent == .chase,
           current.outcome == .swingingStrike {
            return moment(
                context: context,
                tag: .expandAfterTwoStrikes,
                headline: "결정구 유인 성공",
                detail: "2스트라이크 뒤 존 밖으로 유도해 헛스윙 삼진을 만들었습니다."
            )
        }

        if context.balls > context.strikes,
           context.strikes < 2,
           current.intent == .strike,
           addsStrike(current.outcome) {
            return moment(
                context: context,
                tag: .stealStrike,
                headline: "카운트를 되찾았다",
                detail: "타자 우세 카운트에서 스트라이크를 넣어 승부를 원점으로 돌렸습니다."
            )
        }

        guard let previous = recentWindow.last, isValid(previous) else { return nil }

        let velocityDifference = abs(current.expectedVelocityKPH - previous.expectedVelocityKPH)
        if velocityDifference >= minimumSpeedDifferenceKPH,
           disruptsTiming(current.outcome) {
            return moment(
                context: context,
                tag: .speedLadder,
                headline: "속도차 적중 · \(velocityDifference)km/h",
                detail: "앞선 공과 \(velocityDifference)km/h 차이를 만들어 타자의 타이밍을 무너뜨렸습니다."
            )
        }

        if usesOppositeExtremes(previous.zone.row, current.zone.row),
           disruptsTiming(current.outcome) {
            return moment(
                context: context,
                tag: .eyeLevelChange,
                headline: "눈높이를 바꿨다",
                detail: "높은 코스와 낮은 코스를 이어 타자의 시선을 흔들었습니다."
            )
        }

        if usesOppositeExtremes(previous.zone.column, current.zone.column),
           securesResult(current.outcome) {
            return moment(
                context: context,
                tag: .insideOutside,
                headline: "가로 폭을 썼다",
                detail: "몸쪽과 바깥쪽을 연달아 갈라 좋은 결과를 만들었습니다."
            )
        }

        return nil
    }

    private static func recognizesCounterRead(
        context: PlateAppearanceContext,
        current: PitchSequencePitch,
        rivalMemory: RivalMemorySnapshot?
    ) -> Bool {
        guard let rivalMemory, securesResult(current.outcome) else { return false }
        let adaptation = RivalMemoryEngine().analyze(rivalMemory, context: context)
        guard adaptation.detectedPitch != nil || adaptation.detectedZone != nil else {
            return false
        }
        let changedPitch = adaptation.detectedPitch.map { $0 != current.pitchType } ?? false
        let changedZone = adaptation.detectedZone.map { $0 != current.zone } ?? false
        return changedPitch || changedZone
    }

    private static func moment(
        context: PlateAppearanceContext,
        tag: PitchSequenceTag,
        headline: String,
        detail: String
    ) -> PitchSequenceMoment {
        PitchSequenceMoment(
            pitchNumber: context.pitchNumber,
            tag: tag,
            headline: headline,
            detail: detail
        )
    }

    private static func usesOppositeExtremes(_ lhs: Int, _ rhs: Int) -> Bool {
        (lhs == 0 && rhs == 2) || (lhs == 2 && rhs == 0)
    }

    private static func disruptsTiming(_ outcome: PitchOutcome) -> Bool {
        outcome == .swingingStrike || outcome == .inPlayOut
    }

    private static func securesResult(_ outcome: PitchOutcome) -> Bool {
        outcome == .calledStrike || disruptsTiming(outcome)
    }

    private static func addsStrike(_ outcome: PitchOutcome) -> Bool {
        outcome == .calledStrike || outcome == .swingingStrike || outcome == .foul
    }

    private static func isValid(_ pitch: PitchSequencePitch) -> Bool {
        (0...2).contains(pitch.zone.row)
            && (0...2).contains(pitch.zone.column)
            && pitch.expectedVelocityKPH > 0
    }
}

/// Shared economy rule for the non-probabilistic relationship reward.
public enum PitchSequenceMasteryRules {
    public static let maximumTrustReward = 3

    /// Each recognized moment is worth one trust point, with no penalty for absent/invalid data.
    public static func trustReward(for sequenceMasteryCount: Int?) -> Int {
        min(max(sequenceMasteryCount ?? 0, 0), maximumTrustReward)
    }
}

File: apps/android-unity/Assets/Game/Core/Pitching/PitchSequenceEvaluator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Baseball.Core.Domain;

namespace Baseball.Core.Pitching
{
    public enum PitchSequenceTag { SpeedLadder, EyeLevelChange, InsideOutside, ExpandAfterTwoStrikes, StealStrike, CounterRead }

    public sealed class PitchSequencePitch
    {
        public PitchSequencePitch(PitchType pitchType, PitchZone zone, ZoneIntent intent, int expectedVelocityKph, PitchOutcome outcome)
        {
            PitchType = pitchType; Zone = zone; Intent = intent; ExpectedVelocityKph = expectedVelocityKph; Outcome = outcome;
        }
        public PitchType PitchType { get; }
        public PitchZone Zone { get; }
        public ZoneIntent Intent { get; }
        public int ExpectedVelocityKph { get; }
        public PitchOutcome Outcome { get; }
    }

    public sealed class PitchSequenceMoment
    {
        public PitchSequenceMoment(int pitchNumber, PitchSequenceTag tag, string headline, string detail)
        { PitchNumber = pitchNumber; Tag = tag; Headline = headline; Detail = detail; }
        public int PitchNumber { get; }
        public PitchSequenceTag Tag { get; }
        public string Headline { get; }
        public string Detail { get; }
    }

    public static class PitchSequenceEvaluator
    {
        public const int MinimumSpeedDifferenceKph = 12;

        public static PitchSequenceMoment Evaluate(IReadOnlyList<PitchSequencePitch> recent,
            PlateAppearanceContext context, PitchSequencePitch current, RivalMemorySnapshot rivalMemory)
        {
            if (!Valid(current) || context.Balls < 0 || context.Balls > 3 || context.Strikes < 0 || context.Strikes > 2 || context.PitchNumber <= 0)
                return null;
            if (RecognizesCounterRead(context, current, rivalMemory))
                return Moment(context, PitchSequenceTag.CounterRead, "읽힘을 역이용했다", "상대 벤치가 읽은 반복을 끊고 좋은 결과를 만들었습니다.");
            if (context.Strikes == 2 && current.Intent == ZoneIntent.Chase && current.Outcome == PitchOutcome.SwingingStrike)
                return Moment(context, PitchSequenceTag.ExpandAfterTwoStrikes, "결정구 유인 성공", "2스트라이크 뒤 존 밖으로 유도해 헛스윙 삼진을 만들었습니다.");
            if (context.Balls > context.Strikes && context.Strikes < 2 && current.Intent == ZoneIntent.Strike && AddsStrike(current.Outcome))
                return Moment(context, PitchSequenceTag.StealStrike, "카운트를 되찾았다", "타자 우세 카운트에서 스트라이크를 넣어 승부를 원점으로 돌렸습니다.");
            var previous = recent == null ? null : recent.Skip(Math.Max(0, recent.Count - 3)).LastOrDefault();
            if (previous == null || !Valid(previous)) return null;
            var difference = Math.Abs(current.ExpectedVelocityKph - previous.ExpectedVelocityKph);
            if (difference >= MinimumSpeedDifferenceKph && DisruptsTiming(current.Outcome))
                return Moment(context, PitchSequenceTag.SpeedLadder, "속도차 적중 · " + difference + "km/h", "앞선 공과 " + difference + "km/h 차이를 만들어 타자의 타이밍을 무너뜨렸습니다.");
            if (Opposite(previous.Zone.Row, current.Zone.Row) && DisruptsTiming(current.Outcome))
                return Moment(context, PitchSequenceTag.EyeLevelChange, "눈높이를 바꿨다", "높은 코스와 낮은 코스를 이어 타자의 시선을 흔들었습니다.");
            if (Opposite(previous.Zone.Column, current.Zone.Column) && SecuresResult(current.Outcome))
                return Moment(context, PitchSequenceTag.InsideOutside, "가로 폭을 썼다", "몸쪽과 바깥쪽을 연달아 갈라 좋은 결과를 만들었습니다.");
            return null;
        }

        private static bool RecognizesCounterRead(PlateAppearanceContext context, PitchSequencePitch current, RivalMemorySnapshot memory)
        {
            if (memory == null || !SecuresResult(current.Outcome)) return false;
            var adaptation = new RivalMemoryEngine().Analyze(memory, context);
            if (!adaptation.DetectedPitch.HasValue && !adaptation.DetectedZone.HasValue) return false;
            return (adaptation.DetectedPitch.HasValue && adaptation.DetectedPitch.Value != current.PitchType) ||
                   (adaptation.DetectedZone.HasValue && adaptation.DetectedZone.Value != current.Zone);
        }
        private static PitchSequenceMoment Moment(PlateAppearanceContext context, PitchSequenceTag tag, string headline, string detail) =>
            new PitchSequenceMoment(context.PitchNumber, tag, headline, detail);
        private static bool Opposite(int left, int right) => (left == 0 && right == 2) || (left == 2 && right == 0);
        private static bool DisruptsTiming(PitchOutcome outcome) => outcome == PitchOutcome.SwingingStrike || outcome == PitchOutcome.InPlayOut;
        private static bool SecuresResult(PitchOutcome outcome) => outcome == PitchOutcome.CalledStrike || DisruptsTiming(outcome);
        private static bool AddsStrike(PitchOutcome outcome) => outcome == PitchOutcome.CalledStrike || outcome == PitchOutcome.SwingingStrike || outcome == PitchOutcome.Foul;
        private static bool Valid(PitchSequencePitch pitch) => pitch.Zone.Row >= 0 && pitch.Zone.Row <= 2 && pitch.Zone.Column >= 0 && pitch.Zone.Column <= 2 && pitch.ExpectedVelocityKph > 0;
    }

    public static class PitchSequenceMasteryRules
    {
        public const int MaximumTrustReward = 3;
        public static int TrustReward(int? sequenceMasteryCount) => Math.Min(Math.Max(sequenceMasteryCount ?? 0, 0), MaximumTrustReward);
    }
}
